tests/unittests/test_runs/test_fallback_datasource.py

```diff
@@ -0,0 +1,240 @@
+
+BASE_URL = "http://169.254.169.254"
+
+EC2_VERSIONS = [
+    'latest',
+]
+
+
+def fill_properties(props, template=OVF_ENV_CONTENT):
+    lines = []
+    prop_tmpl = '<Property oe:key="{key}" oe:value="{val}"/>'
+    for key, val in props.items():
+        lines.append(prop_tmpl.format(key=key, val=val))
+    indent = "        "
+    properties = ''.join([indent + line + "\n" for line in lines])
+    return template.format(properties=properties)
+
+
+def _register_uris(version, ec2_files, ec2_meta, os_files):
+    """Registers a set of url patterns into httpretty that will mimic the
+    same data returned by the openstack metadata service (and ec2 service)."""
+
+    def match_ec2_url(uri, headers):
+        path = uri.path.strip("/")
+        if len(path) == 0:
+            return (200, headers, "\n".join(EC2_VERSIONS))
+        path = uri.path.lstrip("/")
+        if path in ec2_files:
+            return (200, headers, ec2_files.get(path))
+        if path == 'latest/meta-data/':
+            buf = StringIO()
+            for (k, v) in ec2_meta.items():
+                if isinstance(v, (list, tuple)):
+                    buf.write("%s/" % (k))
+                else:
+                    buf.write("%s" % (k))
+                buf.write("\n")
+            return (200, headers, buf.getvalue())
+        if path.startswith('latest/meta-data/'):
+            value = None
+            pieces = path.split("/")
+            if path.endswith("/"):
+                pieces = pieces[2:-1]
+                value = util.get_cfg_by_path(ec2_meta, pieces)
+            else:
+                pieces = pieces[2:]
+                value = util.get_cfg_by_path(ec2_meta, pieces)
+            if value is not None:
+                return (200, headers, str(value))
+        return (404, headers, '')
+
+    def match_os_uri(uri, headers):
+        path = uri.path.strip("/")
+        if path == 'openstack':
+            return (200, headers, "\n".join([openstack.OS_LATEST]))
+        path = uri.path.lstrip("/")
+        if path in os_files:
+            return (200, headers, os_files.get(path))
+        return (404, headers, '')
+
+    def get_request_callback(method, uri, headers):
+        uri = urlparse(uri)
+        path = uri.path.lstrip("/").split("/")
+        if path[0] == 'openstack':
+            return match_os_uri(uri, headers)
+        return match_ec2_url(uri, headers)
+
+    hp.register_uri(hp.GET, re.compile(r'http://169.254.169.254/.*'),
+                    body=get_request_callback)
+
+
+class TestFallbackDatasource(tests_helpers.FilesystemMockingTestCase,
+                             tests_helpers.HttprettyTestCase):
+    VERSION = 'latest'
+
+    def setUp(self):
+        super(TestFallbackDatasource, self).setUp()
+        self.tmp = self.tmp_dir()
+        self.paths = helpers.Paths(
+            {'cloud_dir': self.tmp, 'run_dir': self.tmp})
+        self.ds = dscd.DataSourceConfigDrive
+
```

Comment:
I think this single class is trying to do too much.  Each of the datasource classes that we want to test correctly fallback  should be their own test class:

TestConfigDriveFallback()
TestOpenStackFallback()
TestOVFFallback()

Then, in the init function for each of these, setup self.ds to the target DS you want to verify.  You can keep all of the common bits in a class, and then subclass the DS that's under test.

```
class TestFallbackDatasource(tests_helpers.FilesystemMockingTestCase,
                             tests_helpers.HttprettyTestCase):
    VERSION = 'latest'

    def setUp(self):
        super(TestFallbackDatasource, self).setUp()
        self.tmp = self.tmp_dir()
        self.paths = helpers.Paths(
            {'cloud_dir': self.tmp, 'run_dir': self.tmp})
...
```

Drop the test_* methods on TestFallbackDatasource and then

```
class TestConfigDriveFallback(TestFallbackDatasource):

    def setUp(self):
            super(TestConfigDriveFallback, self).setUp()
            self.ds = dscd.DataSourceConfigDrive

    def test_invalid_config_drive(self):
...

class TestOpenStackFallback(TestFallbackDatasource):

    def setUp(self):
            super(TestOpenStackFallback, self).setUp()
            self.ds = dsos.DataSourceOpenStack

    def test_invalid_openstack_uri(self):

...

```

Much of the setup in the test_invalid_ methods should be moved to the setUp() method for each of their respective test classes.  This allows additional test cases to reuse the scenario without duplicating the setup.

Reply:
Restructured a bit as per above comments. Let me know if any further changes required. Happy to address